### ISSUE TITLE
Multiple left joins

### ISSUE THREAD

**Issue by mulderr** [tomjaguarpaw/haskell-opaleye#237]
Is there some idiom in Opaleye for composing multiple left joins? Say the desired idealized SQL is:

```sql
select ...
from tabA a
  left join tabB b on (b.aid = a.id)
  left join tabC c on (c.aid = a.id)
  left join tabD d on (d.aid = a.id)
...
```

Best I could find is nesting `leftJoin` which gets unwieldy rather quickly especially that arguments to later `on` clauses get "context sensitive" - there are no table aliases so you basically have to pattern match on the result of the previous left join. N-th left join thus has to pattern match against an N-1 times nested pair to get to the first table.

Above you would get something like `\(((e, _), _), d)` for `d`. Is there currently any way to avoid that?

**Comment by tomjaguarpaw** [comment]
Hi @mulderr, yes I can see this is awkward in Opaleye.

I'm afraid I don't have any good answer for this.  You might find that the new `FunctionalJoin` module gives you some functionality that's a bit easier to work with, but I don't it will be as simple as the plain SQL.

https://www.stackage.org/haddock/nightly-2016-11-09/opaleye-0.5.2.1/Opaleye-FunctionalJoin.html

**Comment by mulderr** [comment]
Thanks, I didn't know about that module. Briefly glancing over the type signatures it looks like the above problem will occur just the same though.

As a semi-related note, it marginally helps readability is to define `leftJoin' c r l = leftJoin l r c`.

**Comment by tomjaguarpaw** [comment]
> As a semi-related note, it marginally helps readability to define leftJoin' c r l = leftJoin l r c.

Yeah, I agree!  Don't want to break the API now though :(  Maybe for the v1 release!

**Comment by tomjaguarpaw** [comment]
> Briefly glancing over the type signatures it looks like the above problem will occur just the same though.

Yeah, it's not *much* better, but it could help a bit.

**Comment by ocharles** [comment]
#265 might be relevant here.

**Comment by saurabhnanda** [comment]
In conjunction with #284 this issue seems to be a major concern when using Opaleye in production. Is there a quick-fix in the form of `leftJoin3`, `leftJoin4`.. `leftJoin10` for 3-way, 4-way, upto 10-way join of tables? (On the lines of various functions that deal with tuples of varying sizes)

**Comment by ocharles** [comment]
I'd try out the above suggestion first and use arrow notation. That's working well for us

**Comment by saurabhnanda** [comment]
@ocharles are you referring to #265 ? Possible to dumb it down for people who don't really understand how `QueryArr` works? Any usage examples of `inlineLeftJoin`?

**Comment by mulderr** [comment]
Closing this one since @ocharles's proposal in #265 seems to work well for my original problem. Discussion continues there.
